packages/angular_devkit/benchmark/src/monitored-process.ts

```diff
@@ -0,0 +1,106 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { SpawnOptions, spawn } from 'child_process';
+import { Observable, Subject, from, timer } from 'rxjs';
+import { concatMap, map, onErrorResumeNext, tap } from 'rxjs/operators';
+import { Command } from './command';
+import { AggregatedProcessStats, MonitoredProcess } from './interfaces';
+const pidusage = require('pidusage');
+const pidtree = require('pidtree');
+const treeKill = require('tree-kill');
+
+
+// Cleanup when the parent process exits.
+const defaultProcessExitCb = () => { };
+let processExitCb = defaultProcessExitCb;
+process.on('exit', () => {
+  processExitCb();
+  processExitCb = defaultProcessExitCb;
+});
+
+export class LocalMonitoredProcess implements MonitoredProcess {
+  private stats = new Subject<AggregatedProcessStats>();
+  private stdout = new Subject<Buffer>();
+  private stderr = new Subject<Buffer>();
+  private pollingRate = 100;
+  stats$: Observable<AggregatedProcessStats> = this.stats.asObservable();
+  stdout$: Observable<Buffer> = this.stdout.asObservable();
+  stderr$: Observable<Buffer> = this.stderr.asObservable();
+
+  constructor(private command: Command) { }
+
+  run(): Observable<number> {
+    return new Observable(obs => {
+      const { cmd, cwd, args } = this.command;
+      const spawnOptions: SpawnOptions = { cwd: cwd, shell: true };
+
+      // Spawn the process.
+      const childProcess = spawn(cmd, args, spawnOptions);
+
+      // Emit output and stats.
+      childProcess.stdout.on('data', (data: Buffer) => this.stdout.next(data));
+      childProcess.stderr.on('data', (data: Buffer) => this.stderr.next(data));
+      const statsSubs = timer(0, this.pollingRate).pipe(
```

Comment:
wow all the rxjs usage here feels un-idiomatic for a command-line tool but I'm learning :)

Reply:
Yeah the three output observables here (`stats$`, `stdout$` and `stderr$`) are not very idiomatic. I didn't quite know how else I could accomplish what I wanted though.

The idea was to allow captures to be setup before the process ran to ensure no output was missed. So the observables had exist when instantiating a `LocalMonitoredProcess`, and output was sent there as it came from the different sources.

I couldn't find a better way to do it. Do you have any suggestions?

Reply:
I was thinking at a higher level, that since our code doesn't need to respond to externally-generated events, we don't really need to be "reactive" in managing state in the app, and could use a more procedural style of programming without any rxjs. It's a broader style discussion for the whole repo.